scalafmt-core/shared/src/main/scala/org/scalafmt/internal/State.scala

```diff
@@ -126,11 +93,66 @@ object State {
       cost + splitWithPenalty.cost,
       // TODO(olafur) expire policy, see #18.
       newPolicy,
-      splits :+ splitWithPenalty,
+      splitWithPenalty,
+      depth + 1,
+      this,
       newIndent,
       newIndents,
       nextStateColumn,
       nextFormatOff
     )
   }
+
+}
+
+object State {
+
+  val start = State(
+    0,
+    PolicySummary.empty,
+    null,
+    0,
+    null,
+    0,
+    Vector.empty[Indent[Num]],
+    0,
+    formatOff = false
+  )
+
+  // this is not best state, it's higher priority for search
+  object Ordering extends Ordering[State] {
+    override def compare(x: State, y: State): Int = compareAt(x, y, 0)
+  }
+
+  private val comparisons: Seq[(State, State) => Int] = Seq(
+    compareCost,
```

Comment:
Also you can encode it without `IndexOutOfBounds` risk

```scala
sealed trait Comparison {
  def compare(s1: State, s2: State): Int
  def next: Option[Comparison]
}
case object Cost extends Comparison {
  override def compare(s1: State, s2: State) = ???
  override def next = Some(SplitsLength)
}
case object SplitsLength extends Comparison {
  override def compare(s1: State, s2: State) = ???
  override def next = Some(SplitOrigin)
}
case object SplitOrigin extends Comparison {
  override def compare(s1: State, s2: State) = ???
  override def next = None
}
```

Reply:
if you feel strongly, i can encode. however, this approach is a bit harder to maintain (reordering or adding a new comparison is not very easy), and is probably a little slower.

i could replace `i < comparisons.length` with the slightly slower
```
  (i + 1) match {
    case comparisons.length => 0
    case next => <tail recursive call>
  }
```
which would be equivalent to `next match { case Some(_) => ...; case None => ... }` in your case.

Reply:
if you feel strongly, i can encode. however, this approach is a bit harder to maintain (reordering or adding a new comparison is not very easy), and is probably a little slower.

i could replace `i < comparisons.length` with the slightly slower
```
  (i + 1) match {
    case comparisons.length => 0
    case next => <tail recursive call>
  }
```
which would be equivalent to `next match { case Some(_) => ...; case None => ... }` needed in your case.

Reply:
My point is ADT encoding protect you from index out of bounds errors in runtime after comparison rules modification. So it simplifying maintaing in this sense. But you are right, there is a trade-off with adding new rules in "linked list".

Anyway it is not a blocker for this PR